packages/syft/src/syft/service/queue/zmq_queue.py

```diff
@@ -285,34 +337,37 @@ def send(self, worker: bytes, message: Union[bytes, List[bytes]]):
     def bind(self, endpoint):
         """Bind producer to endpoint."""
         self.backend.bind(endpoint)
-        print("I: MDP producer/0.1.1 is active at %s", endpoint)
+        logger.info("Producer endpoint: {}", endpoint)
 
     def send_heartbeats(self):
         """Send heartbeats to idle workers if it's time"""
-        if time.time() > self.heartbeat_at:
+        if self.heartbeat_t.has_expired():
             for worker in self.waiting:
                 self.send_to_worker(worker, QueueMsgProtocol.W_HEARTBEAT, None, None)
-            self.heartbeat_at = time.time() + HEARTBEAT_INTERVAL
+            self.heartbeat_t.reset()
 
     def purge_workers(self):
         """Look for & kill expired workers.
 
         Workers are oldest to most recent, so we stop at the first alive worker.
         """
-        while len(self.waiting) > 0:
-            w = self.waiting[0]
-            if w.expiry < time.time():
-                self.delete_worker(w, False)
-                self.waiting.pop(0)
-            else:
-                break
+        # work on a copy of the iterator
```

Comment:
Q: Why do we need to work on a copy of the iterator

Reply:
@rasswanth-s we need to do a copy because we're mutating the original list. If we don't copy we will be mutating while iterating which is bad. Alternate was to use a while loop as used before.

Reply:
yes that loop wasn't cleaning the stale worker completely. found this when i spawned 20 workers and scaled down back to 3.